airflow/providers/mysql/hooks/mysql.py

```diff
@@ -52,7 +52,10 @@ def __init__(self, *args, **kwargs) -> None:
 
     def set_autocommit(self, conn: Connection, autocommit: bool) -> None:  # noqa: D403
         """MySql connection sets autocommit in a different way."""
-        conn.autocommit(autocommit)
+        if hasattr(conn, 'get_autocommit'):  # mysqlclient
+            conn.autocommit(autocommit)
+        else:  # mysql-connector-python
+            conn.autocommit = autocommit
```

Comment:
it's a little awkward to check for attr `get_autocommit` and then not use that attr ...  but rather to only use it as a mechanism to infer that `autocommit` is callable

better would be to just check if autocommit is callable directly:

```suggestion
        if callable(conn.autocommit):
            conn.autocommit(autocommit)
        else:
            conn.autocommit = autocommit
```

Reply:
I avoided doing that because `conn.autocommit` is actually a method with a property decorator around it. It has a side effect of running a `SELECT @@session.autocommit` query whenever you try to access it: https://github.com/mysql/mysql-connector-python/blob/86cceb2bc914362e0990459002976c7e5a271e1d/lib/mysql/connector/connection_cext.py#L148

I also decided against the `callable` function as it seems to not exist in some versions of Python 3. I'm not sure which versions you officially support: https://docs.python.org/3/library/functions.html#callable .

Reply:
All good points lemme take another look

Reply:
This was a tricky one!

I totally agree we should avoid generating the side effect twice.

Here are two options I came up with that I think are a little more direct.

First option is you can check the class of `conn`.  

I am pretty sure that mysql connector will always inherit from this base class:

        if isinstance(conn, MySQLConnectionAbstract):
            conn.autocommit = autocommit
        else:
            conn.autocommit(autocommit)

Though alternatively you could check if it's MySQLdb:

```
from MySQLdb.connections import Connection as MySQLdbConnection
...
if isinstance(conn, MySQLdbConnection):
    conn.autocommit(autocommit)
else:
    conn.autocommit = autocommit
```

And the import alias of MySQLdbConnection can be re-used in resolving that type annotation problem.  I'm not sure what's the best practice re aliasing classes in the event of a collision like we have here but this approach seems reasonable to me.

Option 2 (and I think I like this one) is check if `autocommit` is a property.  You can do so without calling it like so:
```python
        if isinstance(conn.__class__.autocommit, property):
            conn.autocommit = autocommit
        else:
            conn.autocommit(autocommit)
```
I think _also_ checking whether that property has a setter is overkill and not worth the complexity but you could do so with `hasattr(conn.__class__.autocommit, 'fset')`.

But I think I'd go with either of these two approaches because checking for `get_autocommit` and not using it is a little confusing and certainly indirect.

What do you think?

Reply:
Thanks! I like option two because it directly tests the thing it's trying to use.